Add support for libdav1d (AV1 codec)

jaylinski opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

> WORK IN PROGRESS! 

Add support for AV1 codec. (https://en.wikipedia.org/wiki/AV1)

### Tasks

* [x] Linux build [CMake]
* [x] Linux build [Depends]
* [ ] FreeBSD build
* [x] Android build
* [x] macOS build
* [x] iOS build
* [ ] Windows build
* [ ] Debian package build (ffmpeg `autobuild.sh`)
* [ ] Add new deps (meson, ninja-build, pip3) to xbmc-ppa-build-depends
* [x] Add dav1d source to `http://mirrors.kodi.tv/build-deps/sources`

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

AOMedia Video 1 (AV1) is an open, royalty-free video coding format designed for video transmissions over the Internet. It was developed as a successor to VP9 by the Alliance for Open Media (AOMedia), a consortium founded in 2015 that includes semiconductor firms, video on demand providers, video content producers, software development companies and web browser vendors. In Facebook testing that approximates real world conditions AV1 achieved 34%, 46.2% and 50.3% higher data compression than libvpx-vp9, x264 high profile, and x264 main profile respectively.

The main motivation is AV1 support for the Vimeo Plugin, since Vimeo ships AV1 encoded videos.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

A test video from Netfix has been successfully played in Kodi.

Test video source: http://download.opencontent.netflix.com/?prefix=AV1/

## Screenshots (if appropriate):

https://i.imgur.com/XTcfMyC.png

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

## Trying to make sense of the ffmpeg build

There are five (!) different ways ffmpeg is built:

Method | File | Comment
------------ | ------------- | -------------
External | | Enabled via `ENABLE_INTERNAL_FFMPEG OFF` 
CMake | `cmake/modules/FindFFMPEG.cmake` | Uses  `tools/depends/target/ffmpeg/CMakeLists.txt`
Make | `tools/depends/target/ffmpeg/Makefile` | Used by depends build
autobuild.sh | `tools/depends/target/ffmpeg/autobuild.sh` | Used for debian package build
buildffmpeg.sh | `tools/buildsteps/windows/buildffmpeg.sh` | Used for Windows build

jaylinski commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+FROM ubuntu:18.04
```
This is a temporary test file I'm using for compilation-testing.

jaylinski commented on build.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
This is a temporary test file I'm using for compilation-testing.

jaylinski commented on cmake/modules/FindDav1d.cmake:
```diff
@@ -0,0 +1,70 @@
+                            --prefix=${CORE_BUILD_DIR}/dav1d
+                      BUILD_COMMAND ${CMAKE_CURRENT_BINARY_DIR} ninja -C <BINARY_DIR>
+                      INSTALL_COMMAND ${CMAKE_CURRENT_BINARY_DIR} ninja -C <BINARY_DIR> install
+                      BUILD_BYPRODUCTS ${DAV1D_LIBRARY})
```
This is not working. dav1d is using ninja for building. I don't know how to correctly invoke a meson/ninja build via cmake... 

Would be nice if someone could help me out.

lrusak replied:
You should look at how it's done for flatbuffers, crossguid, etc to allow using the `ENABLE_INTERNAL_LIBDAV1D` style logic. See, https://github.com/xbmc/xbmc/blob/master/CMakeLists.txt#L54-L58

Otherwise for the cmake invoking meson/ninja I don't know of any other packages that use that. I think you are mostly correct though by using `externalproject_add`.

jaylinski replied:
Thanks, got it working now. Still trying to figure out how I can tell ffmpeg where to find the installed dav1d-lib.

jaylinski commented on cmake/modules/FindDav1d.cmake:
```diff
@@ -0,0 +1,70 @@
+# required dav1d library versions
+set(REQUIRED_DAV1D_VERSION 0.5)
+
+# TODO Detect existing dav1d
```
I'll add this as soon as the build is working.

jaylinski commented on tools/depends/target/dav1d/Makefile:
```diff
@@ -0,0 +1,39 @@
+include DAV1D-VERSION
```
Under what circumstances are `Makefiles` in the `depends/target` folder even used? :thinking: 

I'm not sure if this is needed.

lrusak replied:
the target depends are use for any platform that needs to build all the depends (basically ever platform other than linux).

jaylinski replied:
Ok, thanks! Will have to test this in my Windows VM then.

jaylinski commented on tools/depends/target/ffmpeg/FFMPEG-VERSION:
```diff
@@ -1,4 +1,4 @@
 LIBNAME=ffmpeg
-BASE_URL=https://github.com/xbmc/FFmpeg
-VERSION=4.0.4-Leia-18.4
+BASE_URL=https://github.com/jaylinski/FFmpeg
```
This has to be changed back to `https://github.com/xbmc/FFmpeg` as soon as an appropriate tag was created.

lrusak commented:
Thanks for the PR!

I think you will need to split some stuff out though. You should put your Dockerfile's into a commit prefixed with [temp] or just drop those files from the PR. You should also split out the required ffmpeg bump into it's own commit so that it's more obvious.

jaylinski committed:
[cmake] ffmpeg: add support for libdav1d

jaylinski commented:
@lrusak Thanks for your review! I created two commits and removed the tmp files.

As soon as I got the build working, I'll ping you again.

wsnipex commented on cmake/modules/FindDav1d.cmake:
```diff
@@ -0,0 +1,74 @@
+                      PATCH_COMMAND curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py &&
+                                    python3 get-pip.py &&
+                                    pip install meson &&
+                                    apt-get --quiet --assume-yes install ninja-build nasm
```
this is not needed(and wrong). We already have meson in depends: https://github.com/xbmc/xbmc/tree/master/tools/depends/native/meson
For non-depends builds (i.e. normal linux build with system libs), meson should already be installed

jaylinski replied:
thanks for the info! I assumed only dependencies listed here would be installed on linux systems:
https://github.com/xbmc/xbmc/blame/18.5-Leia/docs/README.Linux.md#L73

should I add meson and ninja to this doc?

jaylinski replied:
Edit: I think I got it. Kodi will install ninja/meson by itself. So no need to pre-install it.

wsnipex replied:
yes both should be added, as long as external dav1d(via system libs) is not available on many distros.

wsnipex commented on cmake/modules/FindDav1d.cmake:
```diff
@@ -0,0 +1,74 @@
+                                    pip install meson &&
+                                    apt-get --quiet --assume-yes install ninja-build nasm
+                      INSTALL_COMMAND ""
+                      CONFIGURE_COMMAND meson
```
```suggestion
                      CONFIGURE_COMMAND $(NATIVEPREFIX)/bin/meson
```

wsnipex replied:
this is for building with depends(KODI_DEPENDSBUILD=ON), so you'll need to distinguish the configure_command for the other case (just use "meson" there)

jaylinski replied:
Can you tell me how the `NATIVEPREFIX` variable gets populated? It is always empty when I'm building Kodi...

wsnipex replied:
it's only available if you build with our depends system(tools/depends), which compiles all needed libs from source. This is mandatory on most all platforms (android, osx, ios) and optional on linux

jaylinski commented on .travis.yml:
```diff
@@ -73,8 +73,11 @@ install:
     fi
+  # We can't use the `meson` package, since it dav1d requires at least meson >= 0.47.0.
+  # Ubuntu 18.04 still the outdated 0.45.1 version. So we have to install it via pip.
+  - pip3 install meson
```
This will probably need a `python-pip3` in the previous command.

jaylinski commented on tools/depends/target/Makefile:
```diff
@@ -18,7 +18,7 @@ DEPENDS = \
 
 
-FFMPEG_DEPENDS = gnutls
+FFMPEG_DEPENDS = gnutls dav1d
```
Remove this (or make it work).

lrusak commented:
@jaylinski Do you want to pick this up again now that FFmpeg has been bumped to 4.2.2 (which includes libdav1d support) or would you like someone else to pick it up if you don't have time?

jaylinski commented:
> @jaylinski Do you want to pick this up again now that FFmpeg has been bumped to 4.2.2 (which includes libdav1d support) or would you like someone else to pick it up if you don't have time?

Yeah, would be nice if someone with a deep understanding of the Kodi depends build could pick this up. I can't get it to work...

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak committed:
[depends] add nasm

lrusak committed:
[depends] move meson cross file generation to target Makefile and fill in host info

meson c_args (et al) requires the keys to be in a list similar to a python list.
This commit allows us to leverage our native python to allow doing this easily.

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

Rechi commented on tools/depends/native/meson/Makefile:
```diff
@@ -33,3 +33,17 @@ clean:
+	CPU="$(CPU)" \
+	CFLAGS="$(CFLAGS)" \
+	CXXFLAGS="$(CXXFLAGS)" \
+	LDFLAGS="$(LDFLAGS)" \
```
Using host (target) information in build (native) is wrong.

fuzzard replied:
our of curiosity, how would you approach this?

Split the meson cross file generation into a target dependency?

fuzzard replied:
Nevermind this. I've sent lrusak a bash script that generates the meson cross file purely in bash and can be executed as part of configure. No reliance on any native/target dependencies

lrusak replied:
I consider the cross file as part of the meson build system so installing it with the native package is fine by me

fuzzard replied:
The more i think about this, the more i have to agree with Rechi.
Example use case:

I build for ios -> configure -> build native -> build target -> build kodi
I then want to build for a different platform (tvos)
all i should have to do is dist clean target -> configure -> build target -> build kodi.
native hasnt changed, there should be no need to rebuild native for a target change

With this currently, id have to go and rebuild native meson to regenerate the cross which is required due to flag/linker changes between platforms.

Either need to move the cross file generation to a target dependency, or my other option as a possibility. May be other choices as well, but it shouldnt stay in native

lrusak replied:
sure

garbear commented on cmake/modules/FindDav1d.cmake:
```diff
@@ -0,0 +1,91 @@
+#
+# This will define the following variables::
+#
+# DAV1D_FOUND - system has dav1d
```
Also defined:

```
# DAV1D_INCLUDE_DIRS - the dav1d include directories
# DAV1D_LIBRARIES - the dav1d libraries
```

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
![Screen Shot 2020-02-23 at 11 05 39 am](https://user-images.githubusercontent.com/2861319/75101715-c57bdc00-562c-11ea-944d-6709ce02dec0.png)

Runtime tested on macos all ok. Figure id post this here for others to see @lrusak

fuzzard commented on tools/depends/target/dav1d/Makefile:
```diff
@@ -0,0 +1,52 @@
+CONFIGURE = $(NATIVEPREFIX)/bin/python3 $(NATIVEPREFIX)/bin/meson \
+	--buildtype=release \
+	--prefix=$(PREFIX) \
+	--libdir=lib \
```
i dont believe this is required. It will just automatically set to prefix/lib by default

lrusak replied:
libdir is `lib64` on fedora

fuzzard replied:
Is this makefile used on Linux platforms though? I thought they used the cmakelists, and therefore the finddav1d module to add if enable_internal_dav1d is triggered?

lrusak replied:
It can be used on linux platforms yes. Jenkins linux builds use the unified depends system so this is valid in my opinion (moved to the cross file though).

fuzzard replied:
All good then

fuzzard commented on tools/depends/target/dav1d/Makefile:
```diff
@@ -0,0 +1,52 @@
+# configuration settings
+CONFIGURE = $(NATIVEPREFIX)/bin/python3 $(NATIVEPREFIX)/bin/meson \
+	--buildtype=release \
+	--prefix=$(PREFIX) \
```
I think prefix can be moved out to the cross file. Probably best to move all we can to the cross file in case any other dependency ends up using meson at some point.

```
[paths]
prefix = '$PREFIX'
```

reference https://github.com/mesonbuild/meson/blob/master/docs/markdown/Cross-compilation.md

fuzzard commented on tools/depends/target/dav1d/Makefile:
```diff
@@ -0,0 +1,52 @@
+else
+export CC CXX CFLAGS CXXFLAGS
+endif
+export PKG_CONFIG_LIBDIR=$(PREFIX)/lib/pkgconfig
```
bit of a weird reference for now, but according to https://github.com/mesonbuild/meson/blob/master/docs/markdown/Cross-compilation.md

the key quote states

> Since 0.54.0 The pkg_config_libdir property may point to a list of path used internally by Meson to set the PKG_CONFIG_LIBDIR environment variable for pkg-config. This prevents pkg-config from searching cross dependencies in system directories.

the kicker, there is no 0.54.0 tagged release right now, so more a possible todo in the future, to move this over to the cross file as well.

fuzzard commented on tools/depends/target/dav1d/Makefile:
```diff
@@ -0,0 +1,52 @@
+	--prefix=$(PREFIX) \
+	--libdir=lib \
+	--default-library=static \
+	-Denable_asm=true \
```
probably going to have to hide this into a if check.

According to videolan's docs, they support asm for only certain things
https://code.videolan.org/videolan/dav1d#reached

Key take away, 32bit android and ios should be false i think for now. Most of our other platforms should be ok to leave it on.

lrusak replied:
I'd rather leave it enabled for all. libdav1d's meson build will detect if it's possible to use asm or not. see, https://code.videolan.org/videolan/dav1d/blob/master/meson.build#L64-68

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

fuzzard commented:
out of curiosity, should we update this following line now that we build nasm with this?

https://github.com/xbmc/xbmc/blob/b2f476ea6703c0660ac3c90a270e5b108a1fa702/tools/depends/target/config.site.in#L31

fuzzard commented:
And i know you dont want to go this route lrusak, and thats ok, but i figure i'd link to this for prosperity anyway considering it was already created.

https://github.com/fuzzard/xbmc/commit/e419e81f9bd325aa33781002b1b6bb14e971b708

Generation of cross-file.meson within configure

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

fuzzard commented:
Is the tar.gz going to be removed before final merge as well?

lrusak commented:
> Is the tar.gz going to be removed before final merge as well?

yeaaa, it was just a whoopsie on my part

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
@jaylinski thanks for all your contributions to make this happen!

jaylinski commented:
> @jaylinski thanks for all your contributions to make this happen!

Thanks for finishing it! :+1:

Montellese commented:
Maybe I'm too stupid or out of practice but I didn't manage to build `dav1d` through depends on Ubuntu 18.04 natively. I set the `-DENABLE_INTERNAL_DAV1D` option when calling `cmake` but the build failed in `tools/depends/target/dav1d/Makefile`. I then noticed that this `Makefile` looks different than e.g. `tools/depends/target/rapidjson/Makefile` which can also be built internally. I adjusted the `Makefile` for `dav1d`, see c431578849cddac08a71d1e876d24ce521d6e0db. I have no idea if all of that is correct especially the part about `NATIVEPREFIX`.

But that wasn't enough. I also had to make sure the proper version of `meson` is installed. This is not described in `README.Ubuntu.md`. `README.Linux.md` simply mentions that `meson (>= 0.47.0)` needs to be installed but with no instructions on how to do that. I had to do the following:
```bash
sudo apt-get remove meson
sudo pip3 install meson
update-alternatives --install /usr/bin/meson meson /usr/local/bin/meson 50
```

Can someone tell me what I'm doing wrong?

lrusak commented:
@Montellese `ENABLE_INTERNAL_DAV1D` shouldn't actually need to be specified (only in certain circumstances).

In the non depends build (linux typically) the regular cmake build will detect if it needs to build libdav1d (like ffmpeg). In this case you only need to install meson and ninja on your host system.

If you want to do a full depends build (which isn't needed on linux typically) then meson and ninja are built for you as part of the depends native tools. Then libdav1d would be built using these tools.

Montellese commented:
@lrusak you're right. I guess the initial problem was that I didn't have `meson` and `ninja` installed and that the error message just said that `/build` didn't exist.

What is the proper way to install the right `meson` version on Ubuntu 16.04 / 18.04? The version in the `apt` repository is too old.

jaylinski commented:
> @lrusak you're right. I guess the initial problem was that I didn't have `meson` and `ninja` installed and that the error message just said that `/build` didn't exist.
> 
> What is the proper way to install the right `meson` version on Ubuntu 16.04 / 18.04? The version in the `apt` repository is too old.

I mentioned it here: https://github.com/xbmc/xbmc/blob/master/.travis.yml#L78

But it should be also mentioned in the Ubuntu build guide. Or wait for Ubuntu 20.04 (https://packages.ubuntu.com/focal/meson) an make it the only supported Ubuntu version. :wink:

Montellese commented:
I agree with the former and wouldn't count on the latter. We also still support 16.04 so it's a long way to 20.04 which isn't even available ;-)

a1rwulf commented:
We always support 2 versions, so we'll drop 16.04 once 20.04 os available.

jaylinski commented:
@Montellese I'll open a PR later today. :+1:

rmkimathi commented:
AV1 libdav1d (SW) is working very nice on KODI 19.0-ALPHA1. Tested Git:20200424-c4f5c5e4d1 (However media flag for AV1 is missing in file info. Kindly add the Media Flag for AV1). Thanks
